### PR TITLE
Ruff config

### LINKED ISSUE #14736: Create ruff.toml and re-organize the exception codes
This came out of Astropy Coordination Meeting 2023. As `pyproject.toml` contains more things related to build, it is better to declutter it by moving ruff settings to its own `ruff.toml`. While we're at it, we should also organize the codes by logical sections so it is easier for future maintainers (e.g., "never un-ignore these ones"). Then we add @nstarman and @WilliamJamieson as codeowners for the new `ruff.toml` file.

### PR DESCRIPTION
**Pull request by nstarman**
fixes #14736

### CODE REVIEW

**Comment by nstarman** [comment]
First https://github.com/astropy/astropy/pull/14692, then this.

File: docs/conf.py [hunk 1]
```diff
@@ -237,7 +237,7 @@
 # Output file base name for HTML help builder.
 htmlhelp_basename = project + "doc"
 
-# A dictionary of values to pass into the template engine’s context for all pages.
+# A dictionary of values to pass into the template engine`s context for all pages.
```
Comment (mhvk) [inline docs/conf.py:240]: Why this change?

**Review by mhvk — COMMENTED** [review]
As noted in the issue, I'm not so sure this move is really needed, but if all agreed, then of course it is fine. PR itself is OK except for a strange unrelated change that seems to have slipped in.

**Review by WilliamJamieson — CHANGES_REQUESTED** [review]

File: docs/conf.py (hunk 1 above)
Comment (nstarman) [reply docs/conf.py:240]: B/c I accidentally deleted the ignore for RUF00-something and it corrected this. I can revert... or we could just keep it 🤞.

File: .ruff.toml [hunk 2]
```diff
@@ -0,0 +1,373 @@
+target-version = "py38"
+line-length = 88
+select = ["ALL"]
+exclude=[
+    "astropy/extern/*",
+    "*_parsetab.py",
+    "*_lextab.py"
+]
+extend-ignore = [
```
Comment (eerovaher) [inline .ruff.toml:9]: [`extend-ignore` should be replaced with `ignore`](https://beta.ruff.rs/docs/settings/#extend-ignore): 
```suggestion
ignore = [
```

File: .ruff.toml [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+
+    # ---------------------------------------------------------------
+    # TODO
+
+    # flake8-annotations (ANN) : static typing
+    # TODO: revisit this when Astropy implements typing.
+    "ANN",
+
+    # flake8-unused-arguments (ARG)
+    "ARG001",  # unused-function-argument  # 493 violations (2023-05-03)
```
Comment (eerovaher) [inline .ruff.toml:67]: I don't think writing out these numbers adds much value. Writing out the command to obtain these numbers would be much more useful.
```suggestion
    "ARG001",  # unused-function-argument
```

**Review by eerovaher — CHANGES_REQUESTED** [review]
[Ruff documentation says that if it finds `ruff.toml` then it doesn't normally bother reading `pyproject.toml`](https://beta.ruff.rs/docs/configuration/#pyprojecttoml-discovery), but if we add [`extend = "pyproject.toml"`](https://beta.ruff.rs/docs/settings/#extend) to our `ruff.toml` then we can keep the permanent settings there and write out only the TODO items here. Having the permanent and TODO settings in different files would make it very simple to distinguish between the two, for maintainers and for new contributors alike.

File: .ruff.toml (hunk 3 above)
Comment (nstarman) [reply .ruff.toml:67]: The idea for this is to make it easy for maintainers / contributors to find an issue to work on -- essentially helpful triage. I don't think many / any people will run the command umpteen times before finding something suitable to work on.

Comment (eerovaher) [reply .ruff.toml:67]: `ruff --select="ALL" --statistics .` gives much the same information but its output is guaranteed to be up to date with no maintenance burden. Furthermore, some of these rules affect many files and we should encourage addressing them one subpackage at a time. In such cases the numbers listed in the file are not useful at all, but the command can be easily modified, e.g. `ruff --select="ALL" --statistics astropy/time`

File: .ruff.toml [hunk 4]
```diff
@@ -0,0 +1,373 @@
+target-version = "py38"
```
[the remark this answers is not in the document]
Comment (eerovaher) [reply .ruff.toml:1]: Indeed, see https://github.com/astropy/astropy/pull/14286#issuecomment-1448598385. But this information should also be summarized as a comment in the file.

File: docs/conf.py (hunk 1 above)
Comment (mhvk) [reply docs/conf.py:240]: A backtick is just wrong - but the UTF-8 tick is also silly, so maybe you can replace by a regular `'`?

Comment (nstarman) [reply docs/conf.py]: Done!

File: .ruff.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (nstarman) [reply .ruff.toml]: Done.

File: .ruff.toml (hunk 3 above)
Comment (nstarman) [reply .ruff.toml]: Fine enough with me. I worry slightly that this is adding a step, reducing the odds someone volunteers to fix something. So far it's mostly been myself and @WilliamJamieson.

**Review by mhvk — APPROVED** [review]
This looks all good to me, thanks!

**Comment by pllim** [comment]
@eerovaher and @WilliamJamieson , please re-review. Thanks!

File: .ruff.toml [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+    "TRY300",  # Consider `else` block
+    "TRY301",  # raise-within-try
+    "TRY400",  # error-instead-of-exception
+
+    # flake8-quotes (Q)
+    "Q000",  # use double quotes
+]
+unfixable = [
+    "E711"  # NoneComparison. Hard to fix b/c numpy has it's own None.
+]
```
Comment (eerovaher) [inline .ruff.toml:253]: Does this belong here or should it be in `pyproject.toml`?

Comment (eerovaher) [inline .ruff.toml:253]: Does this belong here or should it be in `pyproject.toml`?

Comment (nstarman) [reply .ruff.toml:253]: I would like to fix this eventually.
Or ask ruff if this can be detected with a sufficiently low false-positive rate.

**Review by eerovaher — APPROVED** [review]

**Review by WilliamJamieson — APPROVED** [review]

**Comment by pllim** [comment]
Thanks, all!
